// persistenceonejs/main/clients/client.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PersistenceClient = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _stargate = require("@cosmjs/stargate");

var _cosmwasmStargate = require("@cosmjs/cosmwasm-stargate");

var _tendermintRpc = require("@cosmjs/tendermint-rpc");

var _config = require("../config/config");

var _rpc = require("../proto/cosmwasm/rpc.query");

var _wallet = require("../wallet/wallet");

var PersistenceClient = /*#__PURE__*/function () {
  function PersistenceClient(wasm, core, query, offlineSigner, mnemonic, config) {
    (0, _classCallCheck2["default"])(this, PersistenceClient);
    (0, _defineProperty2["default"])(this, "mnemonic", void 0);
    (0, _defineProperty2["default"])(this, "config", void 0);
    (0, _defineProperty2["default"])(this, "core", void 0);
    (0, _defineProperty2["default"])(this, "wasm", void 0);
    (0, _defineProperty2["default"])(this, "offlineSigner", void 0);
    (0, _defineProperty2["default"])(this, "query", void 0);
    this.core = core;
    this.wasm = wasm;
    this.query = query;
    this.offlineSigner = offlineSigner;
    this.mnemonic = mnemonic;
    this.config = config;
  }

  (0, _createClass2["default"])(PersistenceClient, null, [{
    key: "init",
    value: function () {
      var _init = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(mnemonic, chainConfig, walletOptions) {
        var config, WalletOptions, offlineSigner, core, wasm, tendermintClient, queryClient, rpc, query;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                config = chainConfig || _config.LocalConfig;
                WalletOptions = walletOptions || _config.DefaultWalletOptions; // Get offline signer

                _context.next = 4;
                return _wallet.wallet.setupNodeLocal(config, mnemonic, WalletOptions);

              case 4:
                offlineSigner = _context.sent;
                _context.next = 7;
                return _stargate.SigningStargateClient.connectWithSigner(config.rpc, offlineSigner, {
                  prefix: config.prefix,
                  gasPrice: config.gasPrice
                });

              case 7:
                core = _context.sent;
                _context.next = 10;
                return _cosmwasmStargate.SigningCosmWasmClient.connectWithSigner(config.rpc, offlineSigner, {
                  prefix: config.prefix,
                  gasPrice: config.gasPrice
                });

              case 10:
                wasm = _context.sent;
                _context.next = 13;
                return _tendermintRpc.Tendermint34Client.connect(config.rpc);

              case 13:
                tendermintClient = _context.sent;
                queryClient = new _stargate.QueryClient(tendermintClient);
                _context.next = 17;
                return (0, _stargate.createProtobufRpcClient)(queryClient);

              case 17:
                rpc = _context.sent;
                _context.next = 20;
                return (0, _rpc.createRPCQueryClient)({
                  rpc: rpc
                });

              case 20:
                query = _context.sent;
                return _context.abrupt("return", new PersistenceClient(wasm, core, query, offlineSigner));

              case 22:
              case "end":
                return _context.stop();
            }
          }
        }, _callee);
      }));

      function init(_x, _x2, _x3) {
        return _init.apply(this, arguments);
      }

      return init;
    }()
  }]);
  return PersistenceClient;
}();

exports.PersistenceClient = PersistenceClient;
